staging/src/k8s.io/apiserver/pkg/server/options/audit.go

```diff
@@ -250,44 +272,88 @@ func (o *AuditOptions) AddFlags(fs *pflag.FlagSet) {
 	o.WebhookOptions.AddFlags(fs)
 	o.WebhookOptions.BatchOptions.AddFlags(pluginwebhook.PluginName, fs)
 	o.WebhookOptions.TruncateOptions.AddFlags(pluginwebhook.PluginName, fs)
+	o.DynamicOptions.AddFlags(fs)
 }
 
-func (o *AuditOptions) ApplyTo(c *server.Config) error {
+func (o *AuditOptions) ApplyTo(
+	c *server.Config,
+	kubeClientConfig *restclient.Config,
+	informers informers.SharedInformerFactory,
+	processInfo *ProcessInfo,
+	serviceResolver utilwebhook.ServiceResolver,
+	authInfoResolverWrapper utilwebhook.AuthenticationInfoResolverWrapper,
+) error {
 	if o == nil {
 		return nil
 	}
+	if c == nil {
+		return fmt.Errorf("server config must be non-nil")
+	}
 
 	// Apply advanced options.
 	// 1. Apply generic options.
-	if err := o.applyTo(c); err != nil {
+	checker, err := o.applyTo(*c)
+	if err != nil {
 		return err
 	}
+	c.AuditPolicyChecker = checker
+
+	// 2. Apply log plugin options.
+	logBackend := o.LogOptions.applyTo(*c)
 
-	// 2. Apply plugin options.
-	if err := o.LogOptions.applyTo(c); err != nil {
+	// 3. Apply webhook plugin options
+	webhookBackend, err := o.WebhookOptions.applyTo(*c)
+	if err != nil {
 		return err
 	}
-	if err := o.WebhookOptions.applyTo(c); err != nil {
+
+	groupVersion, err := schema.ParseGroupVersion(o.WebhookOptions.GroupVersionString)
+	if err != nil {
 		return err
 	}
 
-	if c.AuditBackend != nil && c.AuditPolicyChecker == nil {
-		klog.V(2).Info("No audit policy file provided for AdvancedAuditing, no events will be recorded.")
+	// 4. Apply dynamic options.
+	var dynamicBackend audit.Backend
+	if o.DynamicOptions.enabled() {
+		// if dynamic is enabled the webhook and log backends need to be wrapped in an enforced backend with the static policy
+		if webhookBackend != nil {
+			webhookBackend = pluginenforced.NewBackend(webhookBackend, checker)
+		}
+		if logBackend != nil {
+			logBackend = pluginenforced.NewBackend(logBackend, checker)
+		}
+		// create dynamic backend
+		dynamicBackend, checker, err = o.DynamicOptions.applyTo(*c, kubeClientConfig, informers, processInfo, serviceResolver, authInfoResolverWrapper)
+		if err != nil {
+			return err
+		}
+		c.AuditPolicyChecker = checker
+		dynamicBackend = appendBackend(webhookBackend, dynamicBackend)
+		// truncate options wrap both static and dynamic webhook backends
+		dynamicBackend = o.WebhookOptions.TruncateOptions.wrapBackend(dynamicBackend, groupVersion)
+	} else if webhookBackend != nil {
+		dynamicBackend = o.WebhookOptions.TruncateOptions.wrapBackend(webhookBackend, groupVersion)
+	}
+
+	// 5. Join the log backend with the webhooks
+	c.AuditBackend = appendBackend(logBackend, dynamicBackend)
+
+	if c.AuditBackend != nil {
+		klog.V(2).Infof("Using audit backend: %s", c.AuditBackend)
 	}
 	return nil
 }
 
-func (o *AuditOptions) applyTo(c *server.Config) error {
```

Comment:
if we're no longer doing anything with the passed in config, rename the function to something like `buildPolicyChecker` and stop passing in the unused config